### PR TITLE
mds,mon: deprecate CephFS inline_data support

### PR DESCRIPTION
**Pull request by jtlayton**
I recently sent a proposal to the mailing lists to start deprecating inline_data support:

https://www.spinics.net/lists/ceph-users/msg54688.html

TL;DR: I have some concerns about its safety with the kernel client (i.e., I think there are races where it can corrupt data). The kernel doesn't have write support for it yet. We could try to add it, but that's a large task. Given all that, and the fact that it's not widely used, I'm going to propose that we drop support for this.

This PR just adds a hurdle for someone trying to enable it on a new filesystem. At a minimum, I'd like to see this go in for Octopus, with an eye toward ripping it out completely in Q or R release. 

Other things we should to consider:

- a health warning or something when we encounter an existing filesystem that has inline_data enabled
- a conversion utility that can crawl the filesystem, looking for files that are inlined and uninlining them. Not sure there are enough people using it to warrant this, but we could do it if we thought it would be needed

Anything else?

### CODE REVIEW

**Comment by jtlayton** [comment]
FWIW, this removes a lot of the testing for inline_data as well. We may not want to drop that just yet, since we probably do want to know if something in it breaks before we're ready to get rid of it completely. I may reinstate those pieces (with corrected --yes-i-really-really-mean-it flags).

**Comment by jtlayton** [comment]
Went ahead and reinstated the inline_data testing for now.

**Comment by jtlayton** [comment]
Fair enough. I reinstated the description in experimental-features.rst, and added a paragraph that states that it is being deprecated and will likely be removed in Q release.

**Comment by jtlayton** [comment]
Adding cephfs label. While this is technically a change to the mon code, it mostly concerns cephfs.

File: doc/cephfs/experimental-features.rst [hunk 1]
```diff
@@ -23,6 +23,9 @@ failures within it are unlikely to make non-inlined data inaccessible
 Inline data has always been off by default and requires setting
 the ``inline_data`` flag.
 
+Inline data has been decalared deprecated for the Octopus release, and will
```
Comment (gregsfortytwo) [inline doc/cephfs/experimental-features.rst]: s/decalared/declared/

Comment (jtlayton) [reply doc/cephfs/experimental-features.rst]: Doh! Fixed.

**Comment by jtlayton** [comment]
> This makes sense to me.
> 
> Could perhaps try and get whatever telemetry systems we have to include info about inline data on deployed clusters.

That sounds reasonable. What telemetry systems do we have?

I do think we might ought to add a health warning or something though. It'd be nice if, on upgrade, someone with this enabled on an existing filesystem sees a big yellow warning flag pop up so they know they have something that will need to be addressed.

**Comment by jtlayton** [comment]
> It makes me a bit sad to see this go (vs getting fixed), but it's all about where we choose to invest our time--we can't have it all.

Yeah, it's definitely a cool idea.

I made an initial effort at adding write support to the kernel client and to clean up the races, but it turned out to be pretty difficult to get it right. I think to do this correctly, we'd need to refactor a lot of the code -- maybe add some new layer that handles this without needing to involve the higher layer vfs code as much.

I'm not opposed to keeping this around if someone wants to commit to it as a project, but it'll take some time and effort.

**Comment by batrick** [comment]
retest this please

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-08-25_16:39:25-rados-wip-kefu-testing-2019-08-25-2138-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-08-27_00:38:42-rados-wip-kefu-testing-2019-08-26-1528-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-08-27_06:58:14-rados-wip-kefu-testing-2019-08-27-1029-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-08-27_06:45:12-rgw-wip-kefu-testing-2019-08-27-1029-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2019-08-27_08:50:36-rados-wip-kefu-testing-2019-08-27-1029-distro-basic-smithi/

retest this please.

**Comment by jtlayton** [comment]
One more patch. This one just has the MDS log a warning message when it first gets an MDSMap with inline_data enabled. Still considering a health warning too, but I figure it won't hurt to start here.

**Comment by jtlayton** [comment]
...and add a health warning when we encounter a filesystem with this feature enabled. I was able to confirm that I could mute it as well, which should hopefully allay the concern that @gregsfortytwo had.

**Comment by jtlayton** [comment]
Rebased to account for changes to PendingReleaseNotes

**Comment by jtlayton** [comment]
...and rebased again because people keep adding PendingReleaseNotes.

**Comment by jtlayton** [comment]
Rebased again to deal with merge conflicts.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-09-14_01:11:05-rados-wip-kefu-testing-2019-09-14-0046-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-09-15_06:15:53-rados-wip-kefu-testing-2019-09-14-0046-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-09-15_15:37:26-rados-wip-kefu-testing-2019-09-15-1533-distro-basic-mira/

master:

- http://pulpito.ceph.com/pdonnell-2019-09-14_22:40:03-rados-master-distro-basic-smithi/

**Comment by jtlayton** [comment]
> Need to whitelist the new warning in the inline tests:
> 
> http://pulpito.ceph.com/pdonnell-2019-09-15_06:11:06-fs-wip-pdonnell-testing-20190915.030958-distro-basic-smithi/4308559/

Is this patch sufficient to do that?

**Comment by jtlayton** [comment]
Retest this please

**Comment by jtlayton** [comment]
Retest this please

**Comment by jtlayton** [comment]
Retest this please
